kafka-client/src/main/java/dev/responsive/kafka/internal/clients/AsyncStreamsProducer.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.internal.clients;
+
+import dev.responsive.kafka.api.async.internals.AsyncThreadPoolRegistry;
+import java.time.Duration;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.common.errors.ProducerFencedException;
+
+/**
+ * Simple wrapper around the underlying ResponsiveProducer that handles async stuff under EOS.
+ * <p>
+ * The async producer has two jobs: first, it registers and manages the lifecycle of the
+ * async thread pool associated with this StreamThread. Second, it's used to make sure any
+ * pending async records get flushed through the topology before committing a transaction with
+ * those records.
+ * <p>
+ * If ALOS is used, the async producer will not be used at all and the above responsibilities
+ * fall instead to the {@link AsyncStreamsConsumer}
+ */
+public class AsyncStreamsProducer<K, V> extends DelegatingProducer<K, V> {
+
+  private final String streamThreadName;
+  private final AsyncThreadPoolRegistry asyncThreadPoolRegistry;
+  private final Runnable flushAsyncProcessors;
+
+  public AsyncStreamsProducer(
+      final Producer<K, V> delegate,
+      final AsyncThreadPoolRegistry asyncThreadPoolRegistry
+  ) {
+    super(delegate);
+    this.streamThreadName = Thread.currentThread().getName();
+    this.asyncThreadPoolRegistry = asyncThreadPoolRegistry;
+    final var asyncThreadPoolRegistration = asyncThreadPoolRegistry
+        .startNewAsyncThreadPool(streamThreadName);
+    this.flushAsyncProcessors = asyncThreadPoolRegistration::flush;
+  }
+
+  @Override
+  public void commitTransaction() throws ProducerFencedException {
+    flushAsyncProcessors.run();
+    super.commitTransaction();
+  }
+
+  @Override
+  public void close() {
+    // TODO: Producer is closed on reset under EOSv2 :/
```

Comment:
One thing I haven't decided yet is how we want to handle the case of a TaskMigrated/ProducerFenced/etc under EOSv2 in Streams. When these occur we "reinitialize" the StreamsProducer which actually calls close on the old Producer and creates a new one. This feels kind of silly to do for the async thread pool as well, so in theory we could try to avoid closing it and creating a new one somehow. Haven't yet thought through whether there are any correctness implications there (certainly we would need to at least flush/clear any pending work before recycling the async thread pool across a producer restart?)

Reply:
On second thought I think it's clear we do want to just shut down the current async pool and get a new one in case of a TaskMigratedException. So not an issue